cranelift/isle/isle/src/ast.rs

```diff
@@ -419,3 +420,26 @@ pub enum ArgPolarity {
     /// *from* the extractor op.
     Output,
 }
+
+/// An implicit converter, which can appear in extractor or
+/// constructor position. This sets up an ambient automatic type
+/// conversion that is used when needed.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Converter {
+    /// The Rust function name.
+    pub func: Ident,
+    /// The type to convert *from*.
+    pub from_ty: Ident,
+    /// The type to convert *to*.
+    pub to_ty: Ident,
+    /// Infallibility: if an external converter returns `ToType`
+    /// rather than `Option<ToType>` hence can never fail, it is
+    /// declared as such and allows for slightly better code to be
+    /// generated.
+    pub infallible: bool,
+    /// Whether the converter returns a borrowed reference to the
+    /// `to_ty` rather than a newly constructed value.
+    pub borrow: bool,
```

Comment:
It seems strange to me that we need/have this for conversions but not for regular terms. Every place where we want to insert these implicit conversions works right now and doesn't have this `borrow` flag, so I feel like it shouldn't be necessary to add. I must be missing something here.

Reply:
So this became necessary because conversion functions can be used on both the LHS and RHS, and on the LHS we want to deal in borrows as much as possible (as extractors work today) while on the RHS we're usually building entirely new structs. If we were to lower to terms instead then it wouldn't be necessary, as the extractor/constructor codegen already does basically the right thing here.

Reply:
Removed now, with term-based design.